File: 백준/Gold/5569. 출근 경로/출근 경로.py
import sys

input = sys.stdin.readline

MOD = 100000


def checkRange(y, x):
    return 0 <= y < h and 0 <= x < w


def solve(y, x, isVerticalTurn, isHorizontalTurn, isVerticalMove):
    if not checkRange(y, x):
        return 0

    if (y, x) == (h - 1, w - 1):
        return 1

    if dp[y][x][int(isVerticalTurn)][int(isHorizontalTurn)][int(isVerticalMove)] != -1:
        return dp[y][x][int(isVerticalTurn)][int(isHorizontalTurn)][int(isVerticalMove)]

    result = 0

    if not isVerticalTurn and not isHorizontalTurn:
        if isVerticalMove:
            result += solve(y, x + 1, False, False, True)
            result += solve(y + 1, x, False, True, False)
        else:
            result += solve(y, x + 1, True, False, True)
            result += solve(y + 1, x, False, False, False)
    elif not isVerticalTurn and isHorizontalTurn:
        result += solve(y + 1, x, False, False, False)
    elif isVerticalTurn and not isHorizontalTurn:
        result += solve(y, x + 1, False, False, True)

    dp[y][x][int(isVerticalTurn)][int(isHorizontalTurn)][int(isVerticalMove)] = result % MOD
    return dp[y][x][int(isVerticalTurn)][int(isHorizontalTurn)][int(isVerticalMove)]


w, h = map(int, input().split())
dp = [[[[[-1] * 2 for _ in range(2)] for _ in range(2)] for _ in range(w)] for _ in range(h)]
answer = (solve(0, 1, False, False, True) + solve(1, 0, False, False, False)) % MOD

print(answer)

File: 백준/Gold/5569. 출근 경로/출근 경로.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

  static StringBuilder sb = new StringBuilder();
  static int w, h;
  static int[] dx = new int[]{1, 0};
  static int[] dy = new int[]{0, 1};
  static final int MOD = 100000;
  static int[][][][] dp;

  public static void main(String[] args) throws IOException {
    setUp();

    int answer = (solve(1, 2, 0, 0) + solve(2, 1, 0, 1)) % MOD;

    sb.append(answer);
    output();
  }

  private static int solve(int y, int x, int isTurned, int direction) {
    if (y <= 0 || h < y || x <= 0 || w < x) {
      return 0;
    }
    if (y == h && x == w) {
      return 1;
    }
    if (dp[y][x][isTurned][direction] != -1) {
      return dp[y][x][isTurned][direction];
    }

    int result = 0;
    if (isTurned == 1) {
      result = solve(y + dy[direction], x + dx[direction], 0, direction) % MOD;
    } else {
      result = (result + solve(y + dy[direction], x + dx[direction], 0, direction)) % MOD;
      int newDirection = 1 - direction;
      result = (result + solve(y + dy[newDirection], x + dx[newDirection], 1, newDirection)) % MOD;
    }

    dp[y][x][isTurned][direction] = result;
    return result;
  }

  private static void setUp() throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());
    w = Integer.parseInt(st.nextToken());
    h = Integer.parseInt(st.nextToken());
    dp = new int[h + 1][w + 1][2][2];
    for (int i = 0; i < h + 1; i++) {
      for (int j = 0; j < w + 1; j++) {
        for (int k = 0; k < 2; k++) {
          Arrays.fill(dp[i][j][k], -1);
        }
      }
    }
  }

  private static void output() throws IOException {
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    bw.write(sb.toString());
    bw.flush();
    bw.close();
  }
}